dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java

```diff
@@ -67,23 +56,17 @@ private GroupCommParameterProvider(
   }
 
   @Override
-  public synchronized void push(final LayerParameter[] parameterGradients) {
-    // do not store the input if it is not valid
-    if (!(parameterGradients == null || parameterGradients.length == 0)) {
-      parameterGradientList.add(parameterGradients);
-    }
-
-    if (++pushCount >= batchSize) {
-      pushCount = 0;
-
-      try {
-        parameterGradientReduceSender.send(parameterGradientList);
-      } catch (final NetworkException e) {
-        throw new RuntimeException("NetworkException while trying to send reduce", e);
-      } catch (final InterruptedException e) {
-        throw new RuntimeException("InterruptedException while trying to send reduce", e);
+  public synchronized void push(final int batchSize, final LayerParameter[] parameterGradients) {
+    try {
+      if (batchSize == 0 || parameterGradients == null || parameterGradients.length == 0) {
+        parameterGradientReduceSender.send(Collections.<Pair<Integer, LayerParameter[]>>emptyList());
+      } else {
+        parameterGradientReduceSender.send(Collections.singletonList(new Pair<>(batchSize, parameterGradients)));
```

Comment:
I tried to send `Pair` instead of `List<Pair>`. But, I encountered problems about how to define an empty message for synchronization and how a reducer handle this empty message for aggregation.

If we define `<0, null>` as an empty message, we should treat a `null` array at codecs.
If we define `<0, LayerParameter[0]>` as an empty message, there is an inconsistency for the number of layers and the number of layer parameters. (but, we may ignore this inconsistency.)
For `<0, LayerParameter[n]>` with `null` elements, we should treat `null` elements of an array at codecs.

With `List<Pair>`, the empty message is clear and easy to handle it at reducers. If you have a better idea, please let me know :)